terra_notebook_utils/blobstore/gs.py

```diff
@@ -0,0 +1,196 @@
+                    billing_project=self.billing_project)
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+    @property
+    def _gs_bucket(self):
+        if not getattr(self, "_bucket", None):
+            self._bucket = _get_native_bucket(self.bucket_name, self.credentials, self.billing_project)
+        return self._bucket
+
+    def _get_native_blob(self) -> GSNativeBlob:
+        return _get_native_blob(self._gs_bucket, self.key, self.credentials, self.billing_project)
+
+    @property
+    def url(self) -> str:  # type: ignore # TODO https://github.com/python/mypy/issues/4125
+        return f"{GSBlobStore.schema}{self.bucket_name}/{self.key}"
+
+    def get(self) -> bytes:
+        return self._get_native_blob().download_as_bytes(checksum=None)
+
+    def open(self, chunk_size: int=default_chunk_size):
+        return self._get_native_blob().open(chunk_size=chunk_size, mode="rb")
+
+    def put(self, data: bytes):
+        blob = self._gs_bucket.blob(self.key)
+        blob.upload_from_file(io.BytesIO(data))
+
+    def delete(self):
+        self._get_native_blob().delete()
+
+    def copy_from(self, src_blob: "GSBlob", chunk_size: int=default_chunk_size):
+        assert isinstance(src_blob, type(self))
+        if self.url != src_blob.url:
+            if not src_blob._gs_bucket.user_project:
+                # TODO: always use rewrite when it support requester pays buckets
+                dst_gs_blob = self._gs_bucket.blob(self.key)
+                src_gs_blob = src_blob._gs_bucket.blob(src_blob.key)
+                token: Optional[str] = None
+                while True:
+                    try:
+                        resp = dst_gs_blob.rewrite(src_gs_blob, token)
+                    except gcp_exceptions.NotFound:
+                        raise blobstore.BlobNotFoundError(f"Could not find {src_blob.url}")
+                    if resp[0] is None:
+                        break
+                    else:
+                        token = resp[0]
+            else:
+                if src_blob.size() <= chunk_size:
+                    self.put(src_blob.get())
+                else:
+                    with self.multipart_writer() as writer:
+                        for part in src_blob.iter_content():
+                            writer.put_part(part)
+
+    def download(self, path: str):
+        with indirect_open(path) as fh:
+            cs = self.Hash()
+            for part in self.iter_content():
+                fh.write(part.data)
+                cs.update(part.data)
+            assert cs.matches(self.cloud_native_checksum())
+
+    def exists(self) -> bool:
+        blob = self._gs_bucket.blob(self.key)
+        return blob.exists()
+
+    def size(self) -> int:
+        return self._get_native_blob().size
+
+    def cloud_native_checksum(self) -> str:
+        return self._get_native_blob().crc32c
+
+    @property
+    def Hash(self) -> checksum._Hasher:
```

Comment:
Nit: I'd expect this to be called `Hasher` and as a user, if I call `Blob.hash()` to get the actual RC32C in return.

Reply:
Fair point. This will be handled in a future PR.